### PR TITLE
[berkeley] Pickles encoding and verification hardening

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR
* changes the over-the-wire encoding of proofs to disallow the use of features in wrap proofs
* removes some values from the over-the-wire proofs that we recalculate locally to validate anyway
* splits some of the step and wrap circuit types
* adds a validation for chunks to `Pickles.verify`
* adds a validation for feature flags' consistency with evaluations to `Pickles.verify`
* fixes a small error in `Plonk_checks`.

This PR builds on https://github.com/MinaProtocol/mina/pull/13561.

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

File: src/lib/crypto/kimchi_backend/common/plonk_dlog_proof.ml [hunk 1]
```diff
@@ -276,7 +276,7 @@ module Make (Inputs : Inputs_intf) = struct
     ; lookup_aggregation
     ; lookup_table
     ; lookup_sorted =
-        Array.init 5 ~f:(fun i ->
+        Vector.init Nat.N5.n ~f:(fun i ->
```
Comment (rbonichon) [inline src/lib/crypto/kimchi_backend/common/plonk_dlog_proof.ml:279]: Nice change!

File: src/lib/mina_wire_types/pickles/pickles.mli [hunk 2]
```diff
@@ -59,6 +59,62 @@ module Concrete_ : sig
...
+    module Commitments : sig
+      module V1 : sig
+        type t =
+          { w_comm :
+              (Pasta_bindings.Fp.t * Pasta_bindings.Fp.t)
+              Columns_vec.Stable.V1.t
+          ; z_comm : Pasta_bindings.Fp.t * Pasta_bindings.Fp.t
+          ; t_comm :
+              (Pasta_bindings.Fp.t * Pasta_bindings.Fp.t)
+              Quotient_polynomial_vec.Stable.V1.t
```
Comment (rbonichon) [inline src/lib/mina_wire_types/pickles/pickles.mli:77]: Looks like we might want to give a name (or a least an alias just for the interface) to the type `Pasta_bindings.Fp.t * Pasta_bindings.Fp.t`. Would that be a point?

File: src/lib/pickles/verify.ml [hunk 3]
```diff
@@ -49,210 +40,70 @@ let verify_heterogenous (ts : Instance.t list) =
...
-                      ; prev_x_hat = (x_hat1, _) as prev_x_hat
-                   *)
+                 { statement =
+                     { proof_state
+                     ; messages_for_next_step_proof =
+                         { old_bulletproof_challenges; _ }
+                     }
+                     (* TODO
+                        ; prev_x_hat = (x_hat1, _) as prev_x_hat
+                     *)
```
Comment (rbonichon) [inline src/lib/pickles/verify.ml:59]: Since we kept the TODO here, could we elaborate on it?

File: src/lib/pickles/verify.ml [hunk 4]
```diff
@@ -49,210 +40,70 @@ let verify_heterogenous (ts : Instance.t list) =
...
-          }
+        let non_chunking =
+          let single_chunk = ref true in
+          let _unit_evals =
+            Pickles_types.Plonk_types.Evals.map evals.evals.evals
+              ~f:(fun (x, y) ->
+                single_chunk :=
+                  !single_chunk && Array.length x = 1 && Array.length y = 1 )
+          in
+          !single_chunk
```
Comment (rbonichon) [inline src/lib/pickles/verify.ml:72]: Looks like we do not need `single_chunk`. Let me suggest
```suggestion
          let exception Is_chunked in 
           match
            Pickles_types.Plonk_types.Evals.map evals.evals.evals
              ~f:(fun (x, y) ->
                if Array.length x > 1 || Array.length y = 1 then raise Is_chunked )
            with 
            | exception Is_chunked -> false
            | _unit_evals -> true
```

**Review by rbonichon — COMMENTED** [review]
LGTM. Will make a second pass to raise my review to approve.

File: src/lib/pickles/verify.ml (hunk 3 above)
Comment (mrmr1993) [reply src/lib/pickles/verify.ml:59]: It should be deleted

Comment (mrmr1993) [reply src/lib/pickles/verify.ml:59]: It should be deleted

**Comment by dannywillems** [comment]
!ci-build-me

File: src/lib/crypto/kimchi_backend/common/plonk_dlog_proof.ml [hunk 5]
```diff
@@ -365,9 +365,7 @@ module Make (Inputs : Inputs_intf) = struct
     ; rot_selector
     ; lookup_aggregation
     ; lookup_table
-    ; lookup_sorted =
-        Array.init 5 ~f:(fun i ->
-            Option.try_with_join (fun () -> lookup_sorted.(i)) )
+    ; lookup_sorted = Vector.to_array lookup_sorted
```
Comment (querolita) [inline src/lib/crypto/kimchi_backend/common/plonk_dlog_proof.ml:368]: Would it make sense to have both `lookup_sorted` on line 279 have the same type as here? Meaning, either vectors or arrays in both cases.

Comment (querolita) [inline src/lib/crypto/kimchi_backend/common/plonk_dlog_proof.ml:368]: Would it make sense to have both lookup_sorted on line 279 have the same type as here? Meaning, either vectors or arrays in both cases.

File: src/lib/crypto/kimchi_bindings/stubs/src/linearization.rs [hunk 6]
```diff
@@ -1,13 +1,40 @@
...
     linearization::{constraints_expr, linearization_columns},
 };
 
 /// Converts the linearization of the kimchi circuit polynomial into a printable string.
 pub fn linearization_strings<F: ark_ff::PrimeField + ark_ff::SquareRootField>(
+    uses_custom_gates: bool,
 ) -> (String, Vec<(String, String)>) {
-    let evaluated_cols = linearization_columns::<F>(None);
-    let (linearization, _powers_of_alpha) = constraints_expr::<F>(None, true);
+    let features = if uses_custom_gates {
```
Comment (querolita) [inline src/lib/crypto/kimchi_bindings/stubs/src/linearization.rs:14]: I am not sure I follow the logics of this naming. Meaning, if the circuit uses custom gates, then it means that it must have some special features. To me it looks like `uses_custom_gates = false` would have been assigned to `features = None`.

File: src/lib/pickles/step_verifier.ml [hunk 7]
```diff
@@ -344,12 +344,12 @@ struct
         ( 'a
         , Inputs.Impl.Field.t Import.Scalar_challenge.t
         , _ )
-        Types.Step.Proof_state.Deferred_values.Plonk.Minimal.t )
+        Types.Wrap.Proof_state.Deferred_values.Plonk.Minimal.t )
```
Comment (querolita) [inline src/lib/pickles/step_verifier.ml:347]: Oh, can you give more context about why this is now a wrap proof?

File: src/lib/pickles_types/plonk_types.ml [hunk 8]
```diff
@@ -390,6 +391,69 @@ module Evals = struct
...
+    let lookups_per_row_3 = lookups_per_row_4 || f.lookup in
+    let lookups_per_row_2 = lookups_per_row_3 || f.foreign_field_mul in
+    Array.reduce_exn ~f:( && )
+      [| enable_if range_check0_selector f.range_check0
+       ; enable_if range_check1_selector f.range_check1
+       ; enable_if foreign_field_add_selector f.foreign_field_add
+       ; enable_if foreign_field_mul_selector f.foreign_field_mul
+       ; enable_if xor_selector f.xor
+       ; enable_if rot_selector f.rot
+       ; enable_if lookup_aggregation lookups_per_row_2
```
Comment (querolita) [inline src/lib/pickles_types/plonk_types.ml:433]: delete?

File: src/lib/pickles_types/plonk_types.ml [hunk 9]
```diff
@@ -390,6 +391,69 @@ module Evals = struct
...
+       ; enable_if lookup_table lookups_per_row_2
+       ; Vector.foldi lookup_sorted ~init:true ~f:(fun i acc x ->
+             let flag =
+               (* NB: lookups_per_row + 1 in sorted, due to the lookup table. *)
+               match i with
+               | 0 | 1 | 2 ->
+                   lookups_per_row_2
+               | 3 ->
```
Comment (querolita) [inline src/lib/pickles_types/plonk_types.ml:441]: Once the FFMul changes are propagated, then this branch for `lookups_per_row_3` should include all of `| 0 | 1 | 2 | 3 ` , so this is left as a TODO

**Review by querolita — COMMENTED** [review]
Left some comments and a pending TODO, depending on which branch is merged before.

File: src/lib/crypto/kimchi_backend/common/plonk_dlog_proof.ml (hunk 5 above)
Comment (dannywillems) [reply src/lib/crypto/kimchi_backend/common/plonk_dlog_proof.ml:368]: Using vectors is used above for providing information about the length of the arrays to the type system. Here we want to only have runtime information (i.e. the underlying array values) as we want to pass it to the backend. I am in favor of keeping Vector above for type-safeness in the rest of the codebase.

File: src/lib/pickles/step_verifier.ml (hunk 7 above)
Comment (mrmr1993) [reply src/lib/pickles/step_verifier.ml:347]: It was always a wrap proof. The types aliased before; now they don't

File: src/lib/crypto/kimchi_bindings/stubs/src/linearization.rs (hunk 6 above)
Comment (mrmr1993) [reply src/lib/crypto/kimchi_bindings/stubs/src/linearization.rs:14]: Look at the definition of the function. `None` is 'with flags',`Some(flags)` is without flags, with features

**Review by psteckler — APPROVED** [review]
Approving, though I can't comment meaningfully on the crypto changes.

File: src/lib/pickles/verify.ml (hunk 4 above)
Comment (rbonichon) [reply src/lib/pickles/verify.ml]: Done with `Array.length y > 1`

**Comment by mrmr1993** [comment]
!ci-build-me

**Review by rbonichon — APPROVED** [review]

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!approved-for-mainnet

**Comment by mrmr1993** [comment]
!ci-build-me

**Pull request merged**